backend/data_preparation/extractor/fire_extractor.py

```diff
@@ -0,0 +1,203 @@
+    # However, there are records with only 7 consecutive digits because of the carelessness of gov workers
+    RE_EXTRACT_YEAR_ABNORMAL = re.compile(r"\d{7}")
+
+    @staticmethod
+    def _get_year(record: str) -> int:
+        """
+        Gets the year of the record.
+        :param record: the name of the record.
+                e.g. ca_airport_2_20190716_0940_dd83
+        :return: year of this record
+                e.g. 2019
+        """
+        # get the re expression to extract the year
+        re_match_object_for_year = FireExtractor.RE_EXTRACT_YEAR_NORMAL.search(record)
+        if re_match_object_for_year:
+            # when it is normal (8 digits)
+            return int(re_match_object_for_year.group()[:4])
+        else:
+            # when it is abnormal (7 digits)
+            return int(FireExtractor.RE_EXTRACT_YEAR_ABNORMAL.search(record).group()[:4])
+
+    @staticmethod
+    def _get_datetime_before_2016(record: "shapefile._Record") -> datetime:
+        """
+        Gets the datetime of a certain record before 2016, record schema for datetime is: DATE_ and TIME_
+        Sometimes there is only DATE_ but not TIME_
+        :param record: the whole record of fire, shapefile._Record object.
+                    e.g. Record['CA-LNU', 'JZR5', 'Wragg', datetime.date(2015, 8, 2), '0227', 'IR heat perimeter',
+                    'State Agency', 'Y', '', '2015', datetime.date(2015, 8, 2), '', 'CA-LNU-006678', 8049.33]
+        :return:datetime: e.g. 2015/08/02 02:27
+        """
+        return datetime.datetime.strptime("{:%m%d%Y}".format(record["DATE_"]) + record["TIME_"], '%m%d%Y%H%M')
+
+    @staticmethod
+    def _get_datetime_after_2016(record: "shapefile._Record") -> datetime:
+        """
+        Gets the datetime of a certain record after 2016, record schema for datetime is: PERDATTIME or perDatTime
+        :param record: the whole record of fire, shapefile._Record object.
+                    e.g. Record['CA-LNU', 'JZR5', 'Wragg', datetime.date(2015, 8, 2), '0227', 'IR heat perimeter',
+                    'State Agency', 'Y', '', '2015', datetime.date(2015, 8, 2), '', 'CA-LNU-006678', 8049.33]
+        :return:datetime: e.g. 2015/08/02 02:27
+        """
+        datetime_string = record.as_dict().get('PERDATTIME', record.as_dict().get('perDatTime'))
+        datetime_object = (datetime.datetime.strptime(datetime_string, '%m/%d/%Y %I:%M:%S %p')
+                           if len(datetime_string) > 11 else datetime.datetime.strptime(datetime_string, '%m/%d/%Y'))
+        return datetime_object
+
+    def extract(self, path: str, is_sequential: bool, fire_id: int, state: str) -> Dict[str, str]:
+        """
+        Reads a set of records and return the contents.
+        This is a messy function since the data source is dirty.
+        :param path: str, path of the folder
+                e.g. "C:\myResearch\Wildfires\data\\fire-data\ca_trestle_20190605_1200_dd83"
```

Comment:
we are not using absolute path right?

Reply:
It is using abs path, but it will be given by crawler from glob.glob(os.path.join(FIRE_DATA_DIR, "[a-z]*")) so it is not hard coding and can work on different systems